File: src/main/java/graphs/DFS.java
package graphs;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Scanner;

public class DFS {

    private static int cc[];            // component number for each vertex
    private static int current_c = 0;

    static boolean[] visited;
    static ArrayList<Integer>[] graph;

    private static void explore(int v) {
        visited[v] = true;
        cc[v] = current_c;
        for (int to: graph[v]) {
            if (!visited[to])
                explore(to);
        }
    }

    private static void DFS() {
        Arrays.fill(visited, false);
        int n = graph.length;
        cc = new int[n];
        for (int v = 0; v < n; v++) {
            if (!visited[v]) {
                explore(v);
                current_c++;
            }
        }
    }

    public static void main(String[] args) {

        Scanner in = new Scanner(System.in);

        final int n = 4;
        graph = new ArrayList[n];
        for (int i = 0; i < n; i++)
            graph[i] = new ArrayList<>();
        visited = new boolean[n];


        int m = in.nextInt();
        for (int i = 0; i < m; i++) {
            int from = in.nextInt() - 1;
            int to = in.nextInt() - 1;
            graph[from].add(to);
        }

        DFS();

        System.out.println(Arrays.toString(cc));
    }
}

File: src/main/kotlin/graphs/DFS.kt
package graphs

import java.util.*
import kotlin.collections.ArrayList

private lateinit var cc: IntArray                       // номер компоненты связности для каждой вершины
private lateinit var visited: BooleanArray              // помечаем посещенные вершины (можно 0-1, можно раскраской)
private lateinit var graph: Array<ArrayList<Int>>       // граф, заданный списком смежности

private var currentComponent = 0


private fun dfs() {
    val n = graph.size
    for (v in 0 until n) {
        if (!visited[v]) {
            explore(v)
            currentComponent++
        }
    }
}

private fun explore(v: Int) {
    visited[v] = true
    cc[v] = currentComponent
    for (to in graph[v]) {
        if (!visited[to])
            explore(to)
    }
}

fun main(args: Array<String>) = with(Scanner(System.`in`)) {
    val n = nextInt()
    cc = IntArray(n)
    visited = BooleanArray(n)
    graph = Array(n, { ArrayList<Int>() })

    val m = nextInt()
    for (i in 0 until m) {
        val from = nextInt() - 1
        val to = nextInt() - 1
        graph[from].add(to)
        graph[to].add(from)         // для неориентированного графа
    }

    dfs()

    println("Граф: ${graph.toList()}")
    println("Компоненты связности: ${cc.toList()}")
}